### PR TITLE
feat: implement `poppar` intrinsic

### PR DESCRIPTION
**Pull request by parth121101**
fixes : #3665

### CODE REVIEW

File: src/libasr/pass/intrinsic_functions.h [hunk 1]
```diff
@@ -2965,6 +2965,51 @@ namespace Trailz {
...
+        }
+        int result = count%2;
+        return make_ConstantWithType(make_IntegerConstant_t, result, t1, loc);
+    }
+
+    static inline ASR::expr_t* instantiate_Poppar(Allocator &al, const Location &loc,
+        SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+        Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+            declare_basic_variables("_lcompilers_Poppar_" + type_to_str_python(arg_types[0]));
+        fill_func_arg("i", arg_types[0]);
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:3002]: Please document the fortran code here

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -2965,6 +2965,51 @@ namespace Trailz {
...
+        auto result = declare(fn_name, return_type, ReturnVar);
+        ASR::expr_t *func_call_poppar = b.CallIntrinsic(scope, {arg_types[0]}, {args[0]}, return_type, 0, Popcnt::instantiate_Popcnt);
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:3004]: Call `Popcnt` the same way we do `MOD`

File: src/libasr/pass/intrinsic_functions.h [hunk 3]
```diff
@@ -2965,6 +2965,51 @@ namespace Trailz {
...
+        int count = 0;
+        if (val < 0) {
+            count = kind == 4 ? compute_count(mask2, val) : compute_count(mask1, val);
+        } else {
+            while (val) {
+                count += val & 1; 
+                val >>= 1; 
+            }
+        }
+        int result = count%2;
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2994]: I think this can be done in a clean way by calling `popcnt`, I'll prefer to do it that way.

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Left a few comments, I know this can be tedious but let's do it in a right way so that we do not have to look back to this again. 

Note:

You may also refer https://www.geeksforgeeks.org/finding-the-parity-of-a-number-efficiently/ for an efficient compile time evaluation if possible

File: src/lfortran/semantics/ast_common_visitor.h [hunk 4]
```diff
@@ -762,6 +762,7 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
         {"random_number", {IntrinsicSignature({"r"}, 1, 1)}},
         {"mvbits", {IntrinsicSignature({"from", "frompos", "len", "to", "topos"}, 5, 5)}},
         {"popcnt", {IntrinsicSignature({"i"}, 1, 1)}},
+        {"poppar", {IntrinsicSignature({"i"}, 1, 1)}},
```
Comment (HarshitaKalani) [inline src/lfortran/semantics/ast_common_visitor.h:765]: I think, we don't need to define it here if the number of arguments is fixed to 1.
```suggestion
```

File: src/libasr/pass/intrinsic_functions.h [hunk 5]
```diff
@@ -2965,6 +2965,51 @@ namespace Trailz {
...
+        return make_ConstantWithType(make_IntegerConstant_t, result, t1, loc);
+    }
+
+    static inline ASR::expr_t* instantiate_Poppar(Allocator &al, const Location &loc,
+        SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+        Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+            declare_basic_variables("_lcompilers_Poppar_" + type_to_str_python(arg_types[0]));
```
Comment (HarshitaKalani) [inline src/libasr/pass/intrinsic_functions.h:3001]:
```suggestion
            declare_basic_variables("_lcompilers_poppar_" + type_to_str_python(arg_types[0]));
```

**Review by Pranavchiku — APPROVED** [review]
Great, this is good to merge. Thank you!

**Pull request merged**